WalletWasabi/Crypto/Bip322Signature.cs

```diff
@@ -0,0 +1,82 @@
+using System;
+using NBitcoin;
+
+namespace WalletWasabi.Crypto
+{
+	public class Bip322Signature : IBitcoinSerializable
+	{
+		private Script scriptSig = Script.Empty;
+		private WitScript witness = WitScript.Empty; 
+
+		public Bip322Signature(Script scriptSig, WitScript witness)
+		{
+			this.scriptSig = scriptSig;
+			this.witness = witness; 
+		}
+
+		public Bip322Signature()
+		{
+		}
+
+		public Script ScriptSig => scriptSig;
+
+		public WitScript Witness => witness;
+
+		public void ReadWrite(BitcoinStream bitcoinStream)
+		{
+			bitcoinStream.ReadWrite(ref scriptSig);
+			if (bitcoinStream.Serializing)
+			{
+				witness.WriteToStream(bitcoinStream);
+			}
+			else
+			{
+				witness = WitScript.Load(bitcoinStream);
+			}
+		}
+
+		public bool Verify(uint256 hash, Script scriptPubKey)
+		{
+			if (scriptPubKey.IsScriptType(ScriptType.P2WPKH))
+			{
+				if (ScriptSig.Length != 0)
+				{
+					return false;
+				}
+
+				var witnessParameters = PayToWitPubKeyHashTemplate.Instance.ExtractWitScriptParameters(Witness);
+				if (witnessParameters is null)
+				{
+					return false;
+				}
+
+				if (witnessParameters.PublicKey.GetScriptPubKey(ScriptPubKeyType.Segwit) != scriptPubKey)
+				{
+					return false;
+				}
+				
+				// if (witnessParameters.TransactionSignature.SigHash != SigHash.All)
```

Comment:
@onvej-sl why is this check disabled? It seems it should be there.

Reply:
I think it's for the same reason as in https://github.com/trezor/trezor-firmware/blob/master/core/src/apps/bitcoin/ownership.py#L98-L100.